.gitmodules

```diff
@@ -31,3 +31,6 @@
 [submodule "lib/sdl/SDL2_image"]
 	path = lib/sdl/SDL2_image
 	url = https://github.com/libsdl-org/SDL_image.git
+[submodule "lib/usb/ohci_usbhostlib"]
+	path = lib/usb/ohci_usbhostlib
+	url = https://github.com/Ryzee119/ohci_usbhostlib.git
```

Comment:
> This library was developed by OpenNuvoton (https://github.com/OpenNuvoton) who are a large manufacturer of ARM based embedded microcontrollers. This library was intended for use on their range of chips with OHCI hardware. I specifically imported it from https://github.com/OpenNuvoton/N9H30_emWin_NonOS/tree/master/Library/UsbHostLib with some modifications to make it xbox friendly.


With the linux stack we had major issues were our codebase is heavily outdated, leading to bugs that were already fixed upstream. But because nobody did anything for years, it all had a lot of bitrot.
So co-maintaining (or even just using it) is generally better than doing it alone.

- How "bad" would it be to submodule the entire original repository and only use the USB portion?
- Could upstream be convinced to submodule their USB subsystem?
- Are we the only ones who took the submodule? What are the plans: make it highly Xbox specific, or make it re-usable for all sorts of platforms?
- Could the Xbox specific code be moved out-of-tree of the USB stack? How hard would it be (hooks?)

I'd also appreciate if we weren't the only users, so it might be easier to have shared projects, like a libusb for it (does one already exist maybe?).

Reply:
> How "bad" would it be to submodule the entire original repository and only use the USB portion?

Not a huge fan. But can see the benefits. It creates alot of unused code that might confuse people. Afair there's still no good way to submodule a subdirectory?

> Could upstream be convinced to submodule their USB subsystem?

This is ideal, I have created an issue upsteam https://github.com/OpenNuvoton/N9H30_emWin_NonOS/issues/4. They have been quite active so I would expect a response within a few dayts.

> ... highly Xbox specific, or make it re-usable for all sorts of platforms?.
> Could the Xbox specific code be moved out-of-tree of the USB stack? How hard would it be (hooks?)

Being OHCI its very generic already. The only "xbox" specific parts are some kernal calls for memory allocation and interrupt handling. This could be wrapped in a platform/xbox file. Its very platform independant. I dont think it would be hard.

I quick scan of the internet, I can really see any open source projects using this stack.  Its probably used in many thousands of embedded products that we would never know about :)
OHCI is very old, and there is likely more popular EHCI implementations that people would use.

Reply:
> [...]  Afair there's still no good way to submodule a subdirectory?

No, unfortunately there isn't.
We have the same issue with https://github.com/XboxDev/nxdk-libcxx/issues/3; so we might require a solution for this anyway.

Having a kind of "synchronized GitHub mirror (of a subdirectory of monorepo + overlay of other changes)" looks like a useful tech to have, so I'm sure that somewhere, some solution exists (like https://github.com/kubernetes/git-sync + `git filter-branch` hook + generate README which documents what original commit was synchronized last).

But maybe I'm overcomplicating it?

>> How "bad" would it be to submodule the entire original repository and only use the USB portion?
>
> Not a huge fan. But can see the benefits. It creates alot of unused code that might confuse people. [...]

What else is part of upstream? Is it too large like the libcxx repository?

>> Could upstream be convinced to submodule their USB subsystem?
>
> This is ideal, I have created an issue upsteam. They have been quite active so I would expect a response within a few dayts.

Link?

> Being OHCI its very generic already. The only "xbox" specific parts are some kernal calls for memory allocation and interrupt handling. This could be wrapped in a platform/xbox file. Its very platform independant. I dont think it would be hard.

Okay, I'd imagine it like this:
```
lib/usb/libusbohci/             < upstream mirror submodule
lib/usb/libusbohci_xbox/port.c  < only the usb driver
lib/usb/Makefile                < bring upstream + our fork together
```

So we can independently update the upstream and have a clean history for our Xbox stuff.

>  OHCI is very old, and there is likely more popular EHCI implementations that people would use.

Interesting. I didn't even consider that.

So probably not many stakeholders, so might not worth it to put too much effort into making a re-usable fork.
Maybe the ps2dev people (or similar) might still be interested?

Reply:
So update on this side of this. The issue upsteam was raised here regarding a separate usb host lib repo to allow a clean fork https://github.com/OpenNuvoton/N9H30_emWin_NonOS/issues/4

The response wasnt negative but not a clear positive either; (or a timeline), but I'm happy to wait a week or two incase they do this as this is really my preferred solution.

Secondly,

>lib/usb/libusbohci/             < upstream mirror submodule
lib/usb/libusbohci_xbox/port.c  < only the usb driver
lib/usb/Makefile                < bring upstream + our fork together

I like this idea, and will probably make it like this. The "upstream mirror submodule" will likely be my own forked submodule of "upstream" with various commits to make it xbox friendlier but not platform specific.

This will take me a bit of work, so any preferrences or further details on this let me know.

Reply:
I make the neccessary changes so now the repo/submodule cofiguration looks like:

>lib/usb/libusbohci (submoduled lib)
lib/usb/libusbohci_xbox/port.c (Xbox specific functions/config)
lib/usb/Makefile

Reply:
xid xbox specifc driver added to lib/usb/libusbohci_xbox. Upstream HID class is just generic HID class now.

Now all xbox specific parts are in `lib/usb/libusbohci_xbox`.

I think this makes a cleaner way for adding xbox specific drivers in the future (like communicator, camera etc) without polluting the ohci lib too much.